File: backend/controller/task_controller.go
package controller

import (
	"encoding/json"
	"net/http"
	"ptihsan/logs"
	"ptihsan/model"
	"ptihsan/service"
	"ptihsan/utils"
	"strconv"
	"time"
)

func CreateTask(w http.ResponseWriter, r *http.Request) {
	start := time.Now()

	// Ambil input dari body request
	var input struct {
		Title string `json:"title"`
	}
	if err := json.NewDecoder(r.Body).Decode(&input); err != nil {
		logs.LogError(r, "Bad Request: "+err.Error(), http.StatusBadRequest, time.Since(start))
		utils.RespondError(w, http.StatusBadRequest, "Permintaan tidak valid")
		return
	}

	// Buat task baru
	task := model.Task{
		Title:     input.Title,
		Completed: false,
	}

	newTask, err := service.CreateTask(r, task)
	if err != nil {
		logs.LogError(r, "Error creating task: "+err.Error(), http.StatusInternalServerError, time.Since(start))
		utils.RespondError(w, http.StatusInternalServerError, "Gagal membuat task")
		return
	}

	logs.LogInfo(r, "Task created successfully, ID: "+strconv.Itoa(int(newTask.ID)), http.StatusCreated, time.Since(start))
	utils.RespondSuccess(w, http.StatusCreated, "Task berhasil dibuat", newTask)
}

func UpdateTask(w http.ResponseWriter, r *http.Request) {
	start := time.Now()

	// Ambil ID dari query
	id, err := strconv.ParseUint(r.URL.Query().Get("id"), 10, 32)
	if err != nil {
		logs.LogError(r, "Invalid Task ID", http.StatusBadRequest, time.Since(start))
		utils.RespondError(w, http.StatusBadRequest, "Task ID tidak valid")
		return
	}

	// Ambil input title dari body
	var input struct {
		Title string `json:"title"`
	}
	if err := json.NewDecoder(r.Body).Decode(&input); err != nil {
		logs.LogError(r, "Bad Request", http.StatusBadRequest, time.Since(start))
		utils.RespondError(w, http.StatusBadRequest, "Permintaan tidak valid")
		return
	}

	// Update task melalui service
	updatedTask, err := service.UpdateTask(r, uint(id), input.Title)
	if err != nil {
		logs.LogError(r, "Error updating task", http.StatusInternalServerError, time.Since(start))
		utils.RespondError(w, http.StatusInternalServerError, "Gagal mengupdate task")
		return
	}

	logs.LogInfo(r, "Task updated successfully", http.StatusOK, time.Since(start))
	utils.RespondSuccess(w, http.StatusOK, "Task berhasil diupdate", updatedTask)
}

func DeleteTask(w http.ResponseWriter, r *http.Request) {
	start := time.Now()

	// Ambil ID dari query
	id, err := strconv.ParseUint(r.URL.Query().Get("id"), 10, 32)
	if err != nil {
		logs.LogError(r, "Invalid Task ID", http.StatusBadRequest, time.Since(start))
		utils.RespondError(w, http.StatusBadRequest, "Task ID tidak valid")
		return
	}

	// Hapus task via service
	if err := service.DeleteTask(r, uint(id)); err != nil {
		logs.LogError(r, "Gagal menghapus task", http.StatusInternalServerError, time.Since(start))
		utils.RespondError(w, http.StatusInternalServerError, "Gagal menghapus task")
		return
	}

	logs.LogInfo(r, "Task deleted successfully", http.StatusOK, time.Since(start))
	utils.RespondSuccess(w, http.StatusOK, "Task berhasil dihapus", nil)
}

func GetOngoingTasks(w http.ResponseWriter, r *http.Request) {
	start := time.Now()

	tasks, err := service.GetOngoingTasks(r)
	if err != nil || len(tasks) == 0 {
		logs.LogError(r, "Data task ongoing tidak ditemukan", http.StatusNotFound, time.Since(start))
		utils.RespondError(w, http.StatusNotFound, "Data task ongoing tidak ditemukan")
		return
	}

	logs.LogInfo(r, "Fetched ongoing tasks", http.StatusOK, time.Since(start))
	utils.RespondSuccess(w, http.StatusOK, "Data task ongoing ditemukan", tasks)
}

func GetCompletedTasks(w http.ResponseWriter, r *http.Request) {
	start := time.Now()

	tasks, err := service.GetCompletedTasks(r)
	if err != nil {
		logs.LogError(r, "Error fetching completed tasks: "+err.Error(), http.StatusInternalServerError, time.Since(start))
		utils.RespondError(w, http.StatusInternalServerError, "Terjadi kesalahan saat mengambil data task completed")
		return
	}

	if len(tasks) == 0 {
		logs.LogError(r, "Data task completed tidak ditemukan", http.StatusNotFound, time.Since(start))
		utils.RespondError(w, http.StatusNotFound, "Data task completed tidak ditemukan")
		return
	}

	logs.LogInfo(r, "Fetched completed tasks", http.StatusOK, time.Since(start))
	utils.RespondSuccess(w, http.StatusOK, "Data task completed ditemukan", tasks)
}

func SignToCompleted(w http.ResponseWriter, r *http.Request) {
	start := time.Now()

	// Ambil ID dari query
	id, err := strconv.ParseUint(r.URL.Query().Get("id"), 10, 32)
	if err != nil {
		logs.LogError(r, "Invalid Task ID", http.StatusBadRequest, time.Since(start))
		utils.RespondError(w, http.StatusBadRequest, "Task ID tidak valid")
		return
	}

	// Tandai task sebagai completed
	if err := service.MarkTaskAsCompleted(r, uint(id)); err != nil {
		logs.LogError(r, "Gagal menandai task sebagai selesai", http.StatusInternalServerError, time.Since(start))
		utils.RespondError(w, http.StatusInternalServerError, "Gagal menandai task sebagai selesai")
		return
	}

	logs.LogInfo(r, "Task marked as completed", http.StatusOK, time.Since(start))
	utils.RespondSuccess(w, http.StatusOK, "Task berhasil ditandai sebagai selesai", nil)
}

func GetTaskByID(w http.ResponseWriter, r *http.Request) {
	start := time.Now()

	// Ambil ID dari query parameter
	id, err := strconv.ParseUint(r.URL.Query().Get("id"), 10, 32)
	if err != nil {
		logs.LogError(r, "Invalid Task ID", http.StatusBadRequest, time.Since(start))
		utils.RespondError(w, http.StatusBadRequest, "Task ID tidak valid")
		return
	}

	// Ambil task dari service berdasarkan ID
	task, err := service.GetTaskByID(r, uint(id))
	if err != nil {
		logs.LogError(r, "Task not found", http.StatusNotFound, time.Since(start))
		utils.RespondError(w, http.StatusNotFound, "Task tidak ditemukan")
		return
	}

	logs.LogInfo(r, "Fetched task by ID", http.StatusOK, time.Since(start))
	utils.RespondSuccess(w, http.StatusOK, "Task ditemukan", task)
}

File: frontend/app/Http/Controllers/TaskController.php
<?php

namespace App\Http\Controllers;

use Illuminate\Http\Request;
use Illuminate\Support\Facades\Http;

class TaskController extends Controller
{
    public function index()
    {
        // Ambil data dari API
        $response_o = Http::get(env('BACKEND_URL', 'http://localhost:8080').'/tasks-ongoing');
        $response_c = Http::get(env('BACKEND_URL', 'http://localhost:8080').'/tasks-completed');

        // Cek jika responsenya sukses
        $tasks_o = $response_o->successful() && isset($response_o['data']) ? $response_o['data'] : [];
        $tasks_c = $response_c->successful() && isset($response_c['data']) ? $response_c['data'] : [];

        return view('todo', [
            'tasks_o' => $tasks_o,
            'tasks_c' => $tasks_c,
        ]);
    }

    public function store(Request $request)
    {
        // Validasi input
        $request->validate([
            'title' => 'required|string|max:255',
        ]);

        // Kirim data ke backend Go
        $response = Http::post(env('BACKEND_URL', 'http://localhost:8080').'/tasks', [
            'title' => $request->title,
        ]);

        // Cek apakah berhasil
        if ($response->successful() && $response['status'] === 'success') {
            return redirect()->back()->with('success', $response['message'] ?? 'Task berhasil ditambahkan');
        }

        // Jika gagal
        return redirect()->back()->with('error', $response['message'] ?? 'Gagal menambahkan task');
    }

    public function destroy($id)
    {
        // Kirim permintaan DELETE ke backend Go dengan ID di URL
        $response = Http::delete(env('BACKEND_URL', 'http://localhost:8080')."/tasks?id={$id}");

        // Cek apakah permintaan DELETE berhasil
        if ($response->successful() && $response['status'] === 'success') {
            return redirect()->route('tasks.index')->with('success', 'Task berhasil dihapus');
        }

        // Jika gagal
        return redirect()->route('tasks.index')->with('error', 'Gagal menghapus task');
    }

    public function signtocompleted($id)
    {
        // Kirim permintaan PUT ke backend Go dengan ID di URL sebagai query parameter
        $response = Http::put(env('BACKEND_URL', 'http://localhost:8080')."/sign-to-completed?id={$id}", [
            'completed' => true,  // Body raw untuk menandai task sebagai completed
        ]);

        // Cek apakah permintaan PUT berhasil
        if ($response->successful() && $response['status'] === 'success') {
            return redirect()->route('tasks.index')->with('success', 'Task berhasil ditandai sebagai selesai');
        }

        // Jika gagal
        return redirect()->route('tasks.index')->with('error', 'Gagal menandai task sebagai selesai');
    }

    public function edit($id)
    {
        // Mengambil data task berdasarkan ID dari API
        $response = Http::get(env('BACKEND_URL', 'http://localhost:8080')."/tasks-id?id={$id}");

        // Ambil data dari API
        $response_o = Http::get(env('BACKEND_URL', 'http://localhost:8080').'/tasks-ongoing');
        $response_c = Http::get(env('BACKEND_URL', 'http://localhost:8080').'/tasks-completed');

        // Cek jika responsenya sukses
        $tasks_o = $response_o->successful() && isset($response_o['data']) ? $response_o['data'] : [];
        $tasks_c = $response_c->successful() && isset($response_c['data']) ? $response_c['data'] : [];

        // Mengecek apakah task ditemukan
        if ($response->successful()) {
            $task = $response->json()['data']; // Mengambil data task dari response API
            return view('edit', [
                'task' => $task,
                'tasks_o' => $tasks_o,
                'tasks_c' => $tasks_c,
            ]);
        } else {
            return redirect()->route('tasks.index')->with('error', 'Task tidak ditemukan');
        }
    }

    public function update(Request $request, $id)
    {
        // Validasi input
        $request->validate([
            'title' => 'required|string|max:255',
        ]);

        // Ambil data "title" dari input
        $title = $request->input('title');

        // Kirim PUT request ke API eksternal
        $response = Http::put(env('BACKEND_URL', 'http://localhost:8080')."/tasks?id={$id}", [
            'title' => $title
        ]);

        // Cek apakah request API berhasil
        if ($response->successful()) {
            $data = $response->json();

            if ($data['status'] === 'success') {
                // Redirect dengan pesan sukses jika task berhasil diupdate
                return redirect()->route('tasks.edit', $id)->with('success', 'Task berhasil diupdate');
            } else {
                // Redirect dengan pesan error jika task gagal diupdate
                return redirect()->route('tasks.edit', $id)->with('error', 'Task gagal diupdate: ' . $data['message']);
            }
        } else {
            // Redirect dengan pesan error jika API gagal dijangkau
            return redirect()->route('tasks.edit', $id)->with('error', 'Terjadi kesalahan saat menghubungi API.');
        }
    }

}
